src/System.Private.TypeLoader/src/Internal/Reflection/Execution/AssemblyBinderImplementation.Ecma.cs

```diff
@@ -86,9 +90,86 @@ public PEInfo(RuntimeAssemblyName name, MetadataReader reader, PEReader pe)
             }
         }
 
+        partial void BindEcmaFilePath(string assemblyPath, ref AssemblyBindResult bindResult, ref Exception exception, ref bool? result)
+        {
+            // 1. Get a PEReader over a MemoryMappedView of the PE file
+            PEReader pe = OpenPEFile(assemblyPath, out var memoryMappedView);
+            MetadataReader reader = pe.GetMetadataReader();
+
+            // 2. Create AssemblyName from MetadataReader
+            RuntimeAssemblyName runtimeAssemblyName = reader.GetAssemblyDefinition().ToRuntimeAssemblyName(reader);
+
+            lock (s_ecmaLoadedAssemblies)
+            {
+                // 3. Attempt to bind to already loaded assembly
+                if (Bind(runtimeAssemblyName, cacheMissedLookups: false, out bindResult, out exception))
+                {
+                    result = true;
+                    return;
+                }
+                exception = null;
+
+                // 4. If that fails, then add newly created metareader to global cache of loaded modules
+                PEInfo peinfo = new PEInfo(runtimeAssemblyName, reader, pe, memoryMappedView);
+
+                s_ecmaLoadedAssemblies.Add(peinfo);
+                ModuleList moduleList = ModuleList.Instance;
+                ModuleInfo newModuleInfo = new EcmaModuleInfo(moduleList.SystemModule.Handle, pe, reader);
+                moduleList.RegisterModule(newModuleInfo);
+
+                // 5. Then try to load by name again. This load should always succeed
+                if (Bind(runtimeAssemblyName, cacheMissedLookups: true, out bindResult, out exception))
+                {
+                    result = true;
+                    return;
+                }
+
+                result = false;
+                Debug.Assert(exception != null); // We must have an error on load. At this time this could happen due to ambiguous name matching
+            }
+        }
+
+        public static unsafe PEReader OpenPEFile(string filePath, out MemoryMappedViewAccessor mappedViewAccessor)
+        {
+            // System.Reflection.Metadata has heuristic that tries to save virtual address space. This heuristic does not work
+            // well for us since it can make IL access very slow (call to OS for each method IL query). We will map the file
+            // ourselves to get the desired performance characteristics reliably.
+
+            FileStream fileStream = null;
+            MemoryMappedFile mappedFile = null;
+            MemoryMappedViewAccessor accessor = null;
+            try
+            {
+                // Create stream because CreateFromFile(string, ...) uses FileShare.None which is too strict
+                fileStream = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read, 4096, false);
```

Comment:
There is a very similar code here: https://github.com/dotnet/corert/pull/8017/files#diff-a1889bc8eafbf74ba978799be3b7dab7L127 . Would it make sense to unify?

Reply:
That code uses the PEReader stream constructor which [checks if the stream is a FileStream](https://github.com/dotnet/runtime/blob/5c68e94f2e4dcf9fa3accae6b32f3a23f9dc95bd/src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEReader.cs#L179) and then [uses a 16KB treshold to decide if it should be memory mapped or not](https://github.com/dotnet/runtime/blob/5c68e94f2e4dcf9fa3accae6b32f3a23f9dc95bd/src/libraries/System.Reflection.Metadata/src/System/Reflection/Internal/MemoryBlocks/StreamMemoryBlockProvider.cs#L20-L23). The code i've copied always memory maps and has this comment https://github.com/dotnet/corert/blob/c142d456570b105d18bfb92229bddb78e7f4cfd8/src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.cs#L151-L153
which one do you prefer?

Reply:
I like the memory mapping one better.

The 16kB heuristic is good for scenarios that read metadata only and never access IL. It is not the case with JIT.